core/src/main/scala/kafka/log/ProducerStateManager.scala

```diff
@@ -617,8 +615,6 @@ class ProducerStateManager(val topicPartition: TopicPartition,
       // safe to clear the unreplicated transactions
       unreplicatedTxns.clear()
       loadFromSnapshot(logStartOffset, currentTimeMs)
-    } else {
```

Comment:
Just to clarify: since we still remove the snapshots above, the in-memory bookkeeping and the persistent snapshot files could be inconsistent (as we discussed in the ticket). It also means, if the broker has a failure over then we can still see the unknown pid potentially, right?

Reply:
Also it worth updating the javadoc to emphasize the difference of in-memory cache and persistent snapshots in line 601 above.

Reply:
I didn't want to remove the logStartOffset check because I was worried that the snapshots would accumulate. So yes, this means that in the case of a failover or an unclean shutdown, we would lose the state before the expiration time passes.